Phase 3: Check thread arguments

TobiasWrigstad opened the pull request:
Adds a check to thread constructors ensuring that only cowns, immutables, and externally unique regions can be passed in as thread arguments.

There may be other things to check as well here but this is the MVP consistent with the paper. 

*Update* these checks are now only run if the region invariant is enabled. This is probably a design we want to slightly revisit later. 

*Update 2* Clinic stuff is now sorted.

TobiasWrigstad committed:
Check thread arguments

Adds a check to thread constructors ensuring that only
cowns, immutables, and externally unique regions can
be passed in as thread arguments.

TobiasWrigstad committed:
Made pyrona checks conditional on region use

TobiasWrigstad committed:
Clinic

TobiasWrigstad committed:
Clinic fix

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2815,6 +2815,19 @@ builtin_enableinvariant_impl(PyObject *module)
+builtin_is_pyrona_program_impl(PyObject *module)
+/*[clinic end generated code: output=2b6729469da00221 input=d36655a3dc34a881]*/
+{
+    return Py_is_invariant_enabled() ? Py_True : Py_False;
```
Does feel like a bit of abuse of notions.  I think putting a TODO here, se that we don't merge phase 3 without revisiting.  Really I think we want a has used Pyrona, and that is the name of the flag that is used to determine if we should do the region check.

TobiasWrigstad replied:
I fully agree. Will add the TODO as you suggest.

mjp41 reviewed (dismissed):
LGTM

xFrednet commented on Python/bltinmodule.c:
```diff
@@ -2815,6 +2815,19 @@ builtin_enableinvariant_impl(PyObject *module)
+builtin_is_pyrona_program_impl(PyObject *module)
+/*[clinic end generated code: output=2b6729469da00221 input=d36655a3dc34a881]*/
+{
+    return Py_is_invariant_enabled() ? Py_True : Py_False;
```
```suggestion
    return PyBool_FromLong(Py_is_invariant_enabled());
```

xFrednet commented on Lib/threading.py:
```diff
@@ -895,6 +895,24 @@ class is implemented.
 
+        # Only check when a program uses pyrona
+        if is_pyrona_program():
+            def movable(o):
+                return isinstance(o, Region) and not o.is_open()
```
I guess we should talk about the semantics of `is_open()` at one point. In theory, this will always return true, once the WB is fully in place, since the function call of `movable()` and `o.is_open()` both increase the LRC.

Not a blocker for this PR, but something to consider

xFrednet commented on Lib/threading.py:
```diff
@@ -895,6 +895,24 @@ class is implemented.
                     pass
 
+        # Only check when a program uses pyrona
+        if is_pyrona_program():
```
I wonder if we really want to add the restriction on this level. I'm a bit worried that Pyrona unaware libraries might use threads in the background and just break.

Guess this would be a good question for Waldo, asking: "How many libraries actually spawn threads?"

TobiasWrigstad replied:
Not sure how else to do it. It seems hard to mix safe and unsafe use in the same program since we don't have a way of distinguishing between intentionally safe and intentionally unsafe uses.

xFrednet commented on Lib/threading.py:
```diff
@@ -895,6 +895,24 @@ class is implemented.
+
+            for k, v in kwargs.items():
+                if not (isimmutable(v) or isinstance(v, Cown) or movable(v)):
+                    raise RuntimeError(f'thread was passed {k} : {type(v)} -- '
```
It feels like this should be a specific error and not just a generic `RuntimeError`. Not a blocker for this PR, but could you add a TODO for it? Or if you think this should be the final error type I'd be interested in the reasoning :)

TobiasWrigstad replied:
I had similar thoughts. I'll add a TODO and move on!

TobiasWrigstad replied:
Fixed!

xFrednet reviewed (commented):
Changes LGTM, mostly small comments and thoughts :)

TobiasWrigstad committed:
Improved threading check, added technical debt todos

TobiasWrigstad committed:
Added TODO for builtin_is_pyrona_program_impl

TobiasWrigstad committed:
Stupid trailing whitespace

mjp41 commented:
I think for this, we should ultimately have a different thread start function that gives us DRF.

This can initially be just the same as this, and check the arguments.  When we add multi-core, this can then dispatch to a difference core/interpreter, and will have a different local region. 

Then we would get to a point, where all threads on an interpreter are potentially racing, and across interpreters they aren't.  This would give a good migration story.

I think this is fine for now, but we should add comments, that we are going to back out these changes, and add a new API call that code should migrate too.

TobiasWrigstad committed:
Made separate PyronaThread constructor in using.py

TobiasWrigstad committed:
Added comment suggested by @mjp41 and fixed trailing space bug

TobiasWrigstad commented:
> I think for this, we should ultimately have a different thread start function that gives us DRF.

Agreed. I have now created a separate function `PyronaThread` that is defined in our `using.py` library and that starts a DRF thread. The standard `threading.py` library is reverted back to Python 3.12.0.
 
> This can initially be just the same as this, and check the arguments. When we add multi-core, this can then dispatch to a difference core/interpreter, and will have a different local region.
> 
> Then we would get to a point, where all threads on an interpreter are potentially racing, and across interpreters they aren't. This would give a good migration story.

:+1: 

> 
> I think this is fine for now, but we should add comments, that we are going to back out these changes, and add a new API call that code should migrate too.

I've added a comment on `PyronaThread` in `using.py`.

TobiasWrigstad committed:
Fixed import bug

mjp41 commented on Python/bltinmodule.c:
```diff
@@ -2815,6 +2815,25 @@ builtin_enableinvariant_impl(PyObject *module)
     return Py_EnableInvariant();
 }
 
+/*[clinic input]
+is_pyrona_program as builtin_is_pyrona_program
```
I this used anymore?  Should it still be part of this PR?  I think now there is a separate API, we don't need to expose this?

TobiasWrigstad replied:
Good point. I will remove it!

mjp41 commented on Lib/using.py:
```diff
@@ -45,3 +45,57 @@ def decorator(func):
+            return True
+        return False
+
+    if kwargs is None:
+        for a in args:
+            # rc(args) == 3 because we need to know that the args list is moved into the thread too
+            # rc = 3 because:
+            # 1. ref to args in rc
+            # 2. ref to args on this frame
+            # 3. ref to args on the calling frame
+            if not (ok_share(a) or (ok_move(a) and rc(args) == 3)):
+                raise RuntimeError("Thread was passed an object which was neither immutable, a cown, or a unique region")
+            return Thread(group, target, name, args, daemon)
+    else:
+        for k in kwargs:
+            # rc(args) == 3 because we need to know that keyword dict is moved into the thread too
+            # rc = 3 because:
+            # 1. ref to kwargs in rc
+            # 2. ref to kwargs on this frame
+            # 3. ref to kwargs on the calling frame
+            v = kwargs[k]
+            if not (ok_share(v) or (ok_move(v) and rc(kwargs) == 3)):
+                raise RuntimeError("Thread was passed an object which was neither immutable, a cown, or a unique region")
+            return Thread(group, target, name, kwargs, daemon)
```
Would the following be a bit clearer?
```suggestion
    def check(a, args):
      # rc(args) == 3 because we need to know that the args list is moved into the thread too
      # rc = 3 because:
      # 1. ref to args in rc
      # 2. ref to args on this frame
      # 3. ref to args on the calling framedef check(a, args):
      if not ok_share(a) or (ok_move(a) and rc(args) == 3):
        raise RuntimeError("Thread was passed an object which was neither immutable, a cown, or a unique region")
    if kwargs is None:
        for a in args:
            check(a, args)
            return Thread(group, target, name, args, daemon)
    else:
        for k in kwargs:
            check(k, kwargs)
            return Thread(group, target, name, kwargs, daemon)
```

TobiasWrigstad replied:
Indeed -- good refactoring!

TobiasWrigstad committed:
Refactoring of checking logic

Co-authored-by: Matthew Parkinson <mjp41@users.noreply.github.com>

TobiasWrigstad committed:
Improve test and update RC due to refactored logic

TobiasWrigstad committed:
Removed is_pyrona_program builtin

TobiasWrigstad commented:
@mjp41 I've addressed your comments including adding +1 to the RCs due to your refactoring (I recall now that I did not make that refactoring to begin with to lower the RC's by one, but I'd already forgotten and someone will make this improvement in the future and get hit so better bite the bullet now). I think this is ready to go in!

mjp41 reviewed (approved):
LGTM. Thanks for updating the PR
